lib/Target/TfheRust/TfheRustEmitter.cpp

```diff
@@ -31,12 +40,54 @@
 #include "mlir/include/mlir/Support/LogicalResult.h"     // from @llvm-project
 #include "mlir/include/mlir/Tools/mlir-translate/Translation.h"  // from @llvm-project
 
+#define DEBUG_TYPE "tfhe-rust-emitter"
+
 namespace mlir {
 namespace heir {
 namespace tfhe_rust {
 
 namespace {
 
+graph::Graph<Operation *> getGraph(affine::AffineForOp forOp) {
+  graph::Graph<Operation *> graph;
+  auto block = forOp.getBody();
+
+  // Skip if there isn't any apply_lookup_table
+  if (llvm::none_of(block->getOperations(), [&](Operation &op) {
+        return isa<ApplyLookupTableOp>(op);
+      })) {
+    return graph;
+  }
+
+  for (auto &op : block->getOperations()) {
+    if (!isa<ApplyLookupTableOp, AddOp, ScalarLeftShiftOp>(op)) {
+      continue;
+    }
+    graph.addVertex(&op);
+    for (auto operand : op.getOperands()) {
+      auto *definingOp = operand.getDefiningOp();
+      if (!definingOp || definingOp->getBlock() != block ||
+          !isa<ApplyLookupTableOp, AddOp, ScalarLeftShiftOp>(definingOp)) {
+        continue;
+      }
+      graph.addEdge(definingOp, &op);
+    }
+  }
+
+  return graph;
+}
+
+SmallVector<Value> getCiphertextOperands(ValueRange inputs) {
+  SmallVector<Value> vals;
+  for (Value val : inputs) {
+    if (isa<tfhe_rust::EncryptedUInt3Type>(val.getType())) {
```

Comment:
Maybe slap a TODO on here to support all TfheRust encrypted int types.

Reply:
todo slapped!